fog/ledger/enclave/impl/src/oblivious_utils.rs

```diff
@@ -0,0 +1,454 @@
+            ),
+            // Error doesn't beat spent
+            (
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::Spent,
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::KeyImageError,
+                false,
+                "Shard result is error but overwrites spent client!",
+            ),
+            // Unspent doesn't beat error
+            (
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::KeyImageError,
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::NotSpent,
+                false,
+                "Shard result is unspent but overwrites client error!",
+            ),
+            // Unspent doesn't beat spent
+            (
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::Spent,
+                123456,
+                1,
+                123456,
+                TimestampResultCode::TimestampFound,
+                KeyImageResultCode::NotSpent,
+                false,
+                "Shard result is unspent but overwrites spent client!",
+            ),
+        ];
+
+        for (
+            client_key_image,
+            client_spent_at,
+            client_timestamp,
+            client_timestamp_result_code,
+            client_key_image_result_code,
+            shard_key_image,
+            shard_spent_at,
+            shard_timestamp,
+            shard_timestamp_result_code,
+            shard_key_image_result_code,
+            expected_result,
+            panic_message,
+        ) in test_cases.into_iter()
+        {
+            let client_result = KeyImageResult {
+                key_image: client_key_image.into(),
+                spent_at: client_spent_at,
+                timestamp: client_timestamp,
+                timestamp_result_code: client_timestamp_result_code as u32,
+                key_image_result_code: client_key_image_result_code as u32,
+            };
+            let shard_result = KeyImageResult {
+                key_image: shard_key_image.into(),
+                spent_at: shard_spent_at,
+                timestamp: shard_timestamp,
+                timestamp_result_code: shard_timestamp_result_code as u32,
+                key_image_result_code: shard_key_image_result_code as u32,
+            };
+            let result: bool =
+                should_overwrite_key_image_search_result(&client_result, &shard_result).into();
+            assert_eq!(result, expected_result, "{}", panic_message);
+        }
+    }
```

Comment:
:wrench: A couple of flags on this test that makes it hard to parse:
- The tuples with 12 items, generally functions should be kept to 5 or less arguments. I would argue tuples should be kept to 3 or less.
- The for looping over a vector for test cases, that often indicates parameterized test cases, but not always
For this suggestion:
- I created a function `default_client_key_image()` and I updated the `Vec` in `collate_shard_key_image_search_results()` to use it.
- For boolean expected values it's often clearer to have separate test functions instead of passing the expected as one of the parameters
- With dedicated parameterized tests names i think the failure messages can be omitted
- it's a bit easier to see that all 9 permutations  + keys not matching are accounted for (3 permutations were missing from previous)
```suggestion
    use yare::parameterized;

    #[test]
    fn differing_key_images_do_not_update() {
        let client_result = default_client_key_image(1.into());
        let shard_result = default_client_key_image(2.into());

        let result: bool =
            should_overwrite_key_image_search_result(&client_result, &shard_result).into();
        assert_eq!(result, false);
    }

    #[parameterized(
    client_default_shard_is_spent = { KeyImageResultCode::NotSpent, KeyImageResultCode::Spent},
    client_default_shard_is_error = { KeyImageResultCode::NotSpent, KeyImageResultCode::KeyImageError},
    client_error_shard_is_spent = { KeyImageResultCode::KeyImageError, KeyImageResultCode::Spent},
    client_spent_shard_is_spent = { KeyImageResultCode::Spent, KeyImageResultCode::Spent},
    )]
    fn should_update(client_code: KeyImageResultCode, shard_code: KeyImageResultCode) {
        let mut client_result = default_client_key_image(1.into());
        let mut shard_result = client_result.clone();
        client_result.key_image_result_code = client_code as u32;
        shard_result.key_image_result_code = shard_code as u32;

        let result: bool =
            should_overwrite_key_image_search_result(&client_result, &shard_result).into();
        assert_eq!(result, true);
    }

    #[parameterized(
    client_default_shard_is_not_spent = { KeyImageResultCode::NotSpent, KeyImageResultCode::NotSpent},
    client_error_shard_is_not_spent = { KeyImageResultCode::KeyImageError, KeyImageResultCode::NotSpent},
    client_error_shard_is_error = { KeyImageResultCode::KeyImageError, KeyImageResultCode::KeyImageError},
    client_spent_shard_is_not_spent = { KeyImageResultCode::Spent, KeyImageResultCode::NotSpent},
    client_spent_shard_is_error = { KeyImageResultCode::Spent, KeyImageResultCode::KeyImageError},
    )]
    fn should_not_update(client_code: KeyImageResultCode, shard_code: KeyImageResultCode) {
        let mut client_result = default_client_key_image(1.into());
        let mut shard_result = client_result.clone();
        client_result.key_image_result_code = client_code as u32;
        shard_result.key_image_result_code = shard_code as u32;

        let result: bool =
            should_overwrite_key_image_search_result(&client_result, &shard_result).into();
        assert_eq!(result, false);
    }
```